language/move-model/src/model.rs

```diff
@@ -2177,27 +2177,10 @@ impl<'env> NamedConstantEnv<'env> {
 
 /// Represents a type parameter.
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
-pub struct TypeParameter(pub Symbol, pub TypeConstraint);
+pub struct TypeParameter(pub Symbol, pub AbilityConstraint);
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
-pub enum TypeConstraint {
-    None,
-    Copyable,
-    Resource,
-}
-
-impl TypeConstraint {
-    // TODO(tmn) migrate to abilities
-    fn from(abs: AbilitySet) -> Self {
-        if abs.has_key() {
-            TypeConstraint::Resource
-        } else if abs.has_copy() | abs.has_drop() {
-            TypeConstraint::Copyable
-        } else {
-            TypeConstraint::None
-        }
-    }
-}
+pub struct AbilityConstraint(pub AbilitySet);
```

Comment:
I think keeping the name TypeConstraint would have been fine, but I don't work in this code so have at it. 

But conceptually it is still a constraint on the type passed in, so type constraint is fine enough IMO

Reply:
Right, I think as of now, both TypeConstraint and AbilityConstraint might work because AbilityConstraint is currently the only way to constraint a type. I am slightly in favor of using the more direct wording, AbilityConstraint, just in case we may have other constraints on types in the future.